test/fail_compilation/b12647.d

```diff
@@ -0,0 +1,23 @@
+/* TEST_OUTPUT:
+---
+fail_compilation/b12647.d(15): Error: cannot use 'new' in @nogc function 'b12647.test12647_alloc'
+fail_compilation/b12647.d(21): Error: object.Exception is thrown but not caught
+fail_compilation/b12647.d(19): Error: nothrow function 'b12647.test12647_throw' may throw
+---
+*/
+void assertThrown12647(E)(lazy E expression)
+{
+    expression();
+}
+
+void test12647_alloc() pure @safe @nogc
+{
+    auto e = new byte[4];
+    assertThrown12647(0);
+}
+
+void test12647_throw() nothrow pure @safe
+{
+    throw new Exception("");
+    assertThrown12647(0);
```

Comment:
Those tests are not testing `lazy`, because the error happen in the function. I'll craft you some more test cases.